File: frontend/src/types/todo.ts
export type Priority = 'low' | 'medium' | 'high';

export type Todo = {
  id: string;
  title: string;
  due_date: string | null;
  priority: Priority | null;
  label: string | null;
  completed: boolean;
  created_at: string;
  updated_at: string;
};

export type TodoCreate = {
  title: string;
  due_date: string | null;
  priority: Priority | null;
  label: string | null;
};

export type TodoUpdate = TodoCreate & {
  completed: boolean;
};

export type TodoFilters = {
  search: string;
  completed: 'all' | 'active' | 'completed';
  priority: '' | Priority;
};

File: backend/app/models/todo.py
from datetime import date, datetime
import uuid

from sqlalchemy import Boolean, Date, DateTime, String, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Todo(Base):
    __tablename__ = "todos"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    due_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    priority: Mapped[str | None] = mapped_column(String(16), nullable=True)
    label: Mapped[str | None] = mapped_column(String(255), nullable=True)
    completed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
